sklearn/tree/_splitter.pyx

```diff
@@ -434,9 +434,10 @@ cdef class BestSplitter(BaseDenseSplitter):
 
             self.criterion.reset()
             self.criterion.update(best.pos)
-            best.improvement = self.criterion.impurity_improvement(impurity)
             self.criterion.children_impurity(&best.impurity_left,
                                              &best.impurity_right)
+            best.improvement = self.criterion.impurity_improvement(
+                impurity, best.impurity_left, best.impurity_right)
```

Comment:
IIUC, moving the call to `children_impurity` before `impurity_improvement` allow to pass the impurity of the children. Therefore, we avoid these duplicate calculations:

https://github.com/scikit-learn/scikit-learn/blob/395d6c171f6c087875184262c7ba57289e80ba1a/sklearn/tree/_criterion.pyx#L197-L200

Am I right?

Reply:
yes

Reply:
LGTM so!